### PR TITLE
SOLR-17058: Request param to disable distributed stats request at query time

### PR DESCRIPTION
**Pull request by weiwang19**
https://issues.apache.org/jira/browse/SOLR-17058

# Description

When distributed IDF is enabled in solr cloud by adding one of the cache implementations in solrconfig.xml https://solr.apache.org/guide/solr/latest/deployment-guide/solrcloud-distributed-requests.html#distributedidf,  each solr query will incur a distributed shard request to get term statistics.  For queries that does not use distributed IDF information for scoring, the stats request is not necessary.  

# Solution
Hence I propose to add a distrib.statsCache request param so that the distributed stats request can be disabled at query time. 

1. distrib.statsCache defaults to true. When the param is not present, there is no change to current distributed IDF behavior. 
2. When explicitly set distrib.statsCache=false, distributed stats call is disabled for the current query.   

# Tests

Added unit tests related to distrib IDF  and query result key. 

# Checklist

Please review the following and check all that apply:

- [X] I have reviewed the guidelines for [How to Contribute](https://wiki.apache.org/solr/HowToContribute) and my code conforms to the standards described there to the best of my ability.
- [X] I have created a Jira issue and added the issue ID to my pull request title.
- [X] I have given Solr maintainers [access](https://help.github.com/en/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork) to contribute to my PR branch. (optional but recommended)
- [X] I have developed this patch against the `main` branch.
- [X] I have run `./gradlew check`.
- [X] I have added tests for my changes.
- [x] I have added documentation for the [Reference Guide](https://github.com/apache/solr/tree/main/solr/solr-ref-guide)

### CODE REVIEW

File: solr/core/src/test/org/apache/solr/search/stats/TestDistribIDF.java [hunk 1]
```diff
@@ -263,4 +267,51 @@ private void addDocsRandomly() throws IOException, SolrServerException {
...
+    doc.setField("id", "1");
+    doc.setField("cat", "tv");
+    doc.addField(ShardParams._ROUTE_, "a");
+    client.add(COLLECTION, doc);
+
+    doc = new SolrInputDocument();
+    doc.setField("id", "2");
+    doc.setField("cat", "ipad");
+    doc.addField(ShardParams._ROUTE_, "b");
+    client.add(COLLECTION, doc);
```
Comment (mkhludnev) [inline solr/core/src/test/org/apache/solr/search/stats/TestDistribIDF.java:289]: it doesn't even commit. Right?

File: solr/core/src/test/org/apache/solr/search/stats/TestDistribIDF.java [hunk 2]
```diff
@@ -263,4 +267,51 @@ private void addDocsRandomly() throws IOException, SolrServerException {
...
+    track = (NamedList<Object>) rsp.getDebugMap().get("track");
+    assertNotNull(track);
+    assertNull(track.get("PARSE_QUERY"));
+
+    // distributed stats explicitly enabled
+    query.set(CommonParams.DISABLE_DISTRIB_STATS, "false");
+    rsp = client.query(COLLECTION, query);
+    track = (NamedList<Object>) rsp.getDebugMap().get("track");
+    assertNotNull(track);
+    assertNotNull(track.get("PARSE_QUERY"));
```
Comment (mkhludnev) [inline solr/core/src/test/org/apache/solr/search/stats/TestDistribIDF.java:315]: I understand the testing is hard. I barely understand the existing test logic. I think, we can figure out how to extend the existing one.

File: solr/solrj/src/java/org/apache/solr/common/params/CommonParams.java [hunk 3]
```diff
@@ -289,6 +289,14 @@ public static EchoParamStyle get(String v) {
   @Deprecated(since = "9.4")
   String DISABLE_REQUEST_ID = "disableRequestId";
 
+  /**
+   * Parameter to disable the distributed term statistics request for current query when distributed
+   * IDF is enabled in solrconfig
+   *
+   * <p>Defaults to 'false' if not specified
+   */
+  String DISABLE_DISTRIB_STATS = "disableDistribStats";
```
Comment (mkhludnev) [inline solr/solrj/src/java/org/apache/solr/common/params/CommonParams.java:298]: ~~It's time for bikeshedding~~ Giving the existing parameter `distrib.singlePass`, I think we should name it as  `distrib.statsCache=false`. Where `statsCache` repeats the tag in solrconfig.xml.

**Review by mkhludnev — CHANGES_REQUESTED** [review]
@weiwang19 thank you so much for the contribution! I think it's worthwhile to add. I want to share my opinion, it's quite flexible though.

**Comment by mkhludnev** [comment]
pls apply https://github.com/apache/solr/pull/2078/files/449b3ef8704b9ff71b7f2adbedaf8ba49891f781#diff-94c3b5b0568533a1d3768ad04af1bab6d4e3aebdad98a0a51d241e3be9c1b31dR44 trick with nocommit (!) comment  to pass tests here.

File: solr/solr-ref-guide/modules/deployment-guide/pages/solrcloud-distributed-requests.adoc [hunk 4]
```diff
@@ -445,6 +445,21 @@ For example, the following line makes Solr use the `ExactStatsCache` implementat
 ----
 <statsCache class="org.apache.solr.search.stats.ExactStatsCache"/>
 ----
+When `ExactStatsCache`, `ExactSharedStatsCache` or `LRUStatsCache` implementations are enabled in `solrconfig.xml`, distributed calls to fetch global term stats apply to all queries.
+For queries that do not utilize distributed IDF for score calculation, for example
```
Comment (mkhludnev) [inline solr/solr-ref-guide/modules/deployment-guide/pages/solrcloud-distributed-requests.adoc:449]: to be honest, I don't understand the point of this particular sentence. Line 455 and below make sense and clear.

File: solr/solrj/src/java/org/apache/solr/common/params/CommonParams.java (hunk 3 above)
Comment (weiwang19) [reply solr/solrj/src/java/org/apache/solr/common/params/CommonParams.java]: changed to distrib.statsCache

File: solr/core/src/test/org/apache/solr/search/stats/TestDistribIDF.java (hunk 1 above)
Comment (weiwang19) [reply solr/core/src/test/org/apache/solr/search/stats/TestDistribIDF.java:289]: added commit.

File: solr/core/src/test/org/apache/solr/search/stats/TestDistribIDF.java (hunk 2 above)
Comment (weiwang19) [reply solr/core/src/test/org/apache/solr/search/stats/TestDistribIDF.java:322]: adjusted test query

File: solr/solr-ref-guide/modules/deployment-guide/pages/solrcloud-distributed-requests.adoc (hunk 4 above)
Comment (weiwang19) [reply solr/solr-ref-guide/modules/deployment-guide/pages/solrcloud-distributed-requests.adoc]: Thanks Mikhail. Changed wording to make it more clear.

**Comment by mkhludnev** [comment]
Thanks @weiwang19
I'll check it next week. 
@devs please chime in  - I always hesitate to come up with names in API!

**Comment by weiwang19** [comment]
please let me know if there are any concerns to address.

File: solr/core/src/test/org/apache/solr/search/stats/TestDistribIDF.java (hunk 2 above)
Comment (mkhludnev) [reply solr/core/src/test/org/apache/solr/search/stats/TestDistribIDF.java:322]: Pardon. I don't see actual testing/asserting logic/clue. How we can make sure the this flag changes anything? May we just put a few docs unevenly and assert that score is different with invoking this flag true and false?

**Comment by mkhludnev** [comment]
> please let me know if there are any concerns to address.

pls check https://github.com/weiwang19/solr/pull/1

**Comment by weiwang19** [comment]
Done with the changes.  Thank you so much @mkhludnev

**Review by mkhludnev — APPROVED** [review]

**Comment by mkhludnev** [comment]
@weiwang19 please align the code 
https://github.com/apache/solr/actions/runs/7984305841/job/21803474034?pr=2046
NB: it would be easy if you [ ] allow pushes from maintainers. At least it's available for new PRs.

**Comment by mkhludnev** [comment]
Thanks, @weiwang19. Would you mind to create cherrypick PR to branch_9x?

**Comment by weiwang19** [comment]
@mkhludnev can you help check the PR for backport to 9x?  https://github.com/apache/solr/pull/2291

**Pull request merged**